--- mnesia_log.erl.orig	2009-01-14 13:22:53.000000000 -0800
+++ mnesia_log.erl	2009-01-14 13:45:11.000000000 -0800
@@ -220,6 +220,7 @@
 %% Write commit records to the latest_log
 log(C) when  C#commit.disc_copies == [],
              C#commit.disc_only_copies  == [],
+             C#commit.external_copies  == [],
              C#commit.schema_ops == [] ->
     ignore;
 log(C) ->
@@ -243,6 +244,7 @@
 
 slog(C) when  C#commit.disc_copies == [],
              C#commit.disc_only_copies  == [],
+             C#commit.external_copies  == [],
              C#commit.schema_ops == [] ->
     ignore;
 slog(C) ->
