--- mnesia.app.orig	2009-01-14 13:48:46.000000000 -0800
+++ mnesia.app	2009-01-14 13:49:08.000000000 -0800
@@ -1,6 +1,6 @@
 {application, mnesia,
  [{description, "MNESIA  CXC 138 12"},
-  {vsn, "4.4.7"},
+  {vsn, "4.4.7ext4"},
   {modules, [
 	     mnesia, 
 	     mnesia_backup, 
@@ -10,6 +10,7 @@
 	     mnesia_controller,
 	     mnesia_dumper, 
 	     mnesia_event, 
+	     mnesia_ext, 
 	     mnesia_frag, 
 	     mnesia_frag_hash, 
              mnesia_frag_old_hash,
