--- mnesia-4.4.2/src/mnesia_monitor.erl	2007-11-26 06:07:43.000000000 -0700
+++ mnesia-4.4.2ext1/src/mnesia_monitor.erl	2008-07-11 20:19:41.000000000 -0600
@@ -1,3 +1,5 @@
+%% @hidden
+
 %% ``The contents of this file are subject to the Erlang Public License,
 %% Version 1.1, (the "License"); you may not use this file except in
 %% compliance with the License. You should have received a copy of the
@@ -28,6 +30,7 @@
 	 init/0,
 	 mktab/2,
 	 unsafe_mktab/2,
+         unsafe_create_external/3,
 	 mnesia_down/2,
 	 needs_protocol_conversion/1,
 	 negotiate_protocol/1,
@@ -119,6 +122,8 @@
 unsafe_close_log(Name) ->
     unsafe_call({unsafe_close_log, Name}).
 
+unsafe_create_external(Tab, Mod, Cs) ->
+    unsafe_call({unsafe_create_external, Tab, Mod, Cs}).
 
 disconnect(Node) ->
     cast({disconnect, Node}).
@@ -397,6 +402,14 @@
     disk_log:close(Name),
     {reply, ok, State};
 
+handle_call({unsafe_create_external, Tab, Mod, Cs}, _From, State) ->
+    case catch Mod:create_table(Tab, Cs) of
+	{'EXIT', ExitReason} ->
+	    {reply, {error, ExitReason}, State};
+	Reply ->
+	    {reply, Reply, State}
+    end;
+
 handle_call({negotiate_protocol, Mon, _Version, _Protocols}, _From, State) 
   when State#state.tm_started == false ->
     State2 =  State#state{early_connects = [node(Mon) | State#state.early_connects]},    
