--- /usr/lib/erlang/lib/mnesia-4.3.5/src/mnesia_tm.erl	2007-08-05 08:49:02.000000000 -0700
+++ ./mnesia_tm.erl	2008-06-13 21:08:36.000000000 -0700
@@ -1151,6 +1151,7 @@
        ram_copies       =  lists:reverse(H#commit.ram_copies),
        disc_copies      =  lists:reverse(H#commit.disc_copies),
        disc_only_copies =  lists:reverse(H#commit.disc_only_copies),
+       external_copies =  lists:reverse(H#commit.external_copies),
        snmp             = lists:reverse(H#commit.snmp)
       }  
      | reverse(R)].
@@ -1296,7 +1297,10 @@
 		Rec#commit{disc_copies = [Item | Rec#commit.disc_copies]};
 	    disc_only_copies ->
 		Rec#commit{disc_only_copies =
-			   [Item | Rec#commit.disc_only_copies]}
+			   [Item | Rec#commit.disc_only_copies]};
+	    {external_copies, _} ->
+		Rec#commit{external_copies =
+			   [Item | Rec#commit.external_copies]}
 	end,
     prepare_node(Node, Storage, Items, Rec2, Kind);
 prepare_node(_Node, _Storage, Items, Rec, Kind) 
@@ -1728,7 +1732,8 @@
     R  = do_snmp(Tid, C#commit.snmp),
     R2 = do_update(Tid, ram_copies, C#commit.ram_copies, R),
     R3 = do_update(Tid, disc_copies, C#commit.disc_copies, R2),
-    do_update(Tid, disc_only_copies, C#commit.disc_only_copies, R3).
+    R4 = do_update(Tid, disc_only_copies, C#commit.disc_only_copies, R3),
+    do_update(Tid, external_copies, C#commit.external_copies, R4).
 
 %% Update the items
 do_update(Tid, Storage, [Op | Ops], OldRes) ->
@@ -1871,6 +1876,7 @@
 commit_nodes([C | Tail], AccD, AccR) 
         when C#commit.disc_copies == [],
              C#commit.disc_only_copies  == [],
+             C#commit.external_copies  == [],
              C#commit.schema_ops == [] ->
     commit_nodes(Tail, AccD, [C#commit.node | AccR]);
 commit_nodes([C | Tail], AccD, AccR) ->
@@ -1883,7 +1889,8 @@
     {D2, Tail2} = 
 	case C#commit.schema_ops of
 	    [] when C#commit.disc_copies == [],
-		    C#commit.disc_only_copies  == [] ->
+		    C#commit.disc_only_copies  == [],
+                    C#commit.external_copies  == [] ->
 		commit_decision(D, Tail, AccD, [N | AccR]);
 	    [] ->
 		commit_decision(D, Tail, [N | AccD], AccR);
