--- mnesia-4.4.2/src/mnesia_tm.erl	2008-03-11 06:29:32.000000000 -0600
+++ mnesia-4.4.2ext1/src/mnesia_tm.erl	2008-07-11 20:22:18.000000000 -0600
@@ -1,3 +1,5 @@
+%% @hidden
+
 %% ``The contents of this file are subject to the Erlang Public License,
 %% Version 1.1, (the "License"); you may not use this file except in
 %% compliance with the License. You should have received a copy of the
@@ -1151,6 +1153,7 @@
        ram_copies       =  lists:reverse(H#commit.ram_copies),
        disc_copies      =  lists:reverse(H#commit.disc_copies),
        disc_only_copies =  lists:reverse(H#commit.disc_only_copies),
+       external_copies =  lists:reverse(H#commit.external_copies),
        snmp             = lists:reverse(H#commit.snmp)
       }  
      | reverse(R)].
@@ -1296,7 +1299,10 @@
 		Rec#commit{disc_copies = [Item | Rec#commit.disc_copies]};
 	    disc_only_copies ->
 		Rec#commit{disc_only_copies =
-			   [Item | Rec#commit.disc_only_copies]}
+			   [Item | Rec#commit.disc_only_copies]};
+	    {external_copies, _} ->
+		Rec#commit{external_copies =
+			   [Item | Rec#commit.external_copies]}
 	end,
     prepare_node(Node, Storage, Items, Rec2, Kind);
 prepare_node(_Node, _Storage, Items, Rec, Kind) 
@@ -1728,7 +1734,8 @@
     R  = do_snmp(Tid, C#commit.snmp),
     R2 = do_update(Tid, ram_copies, C#commit.ram_copies, R),
     R3 = do_update(Tid, disc_copies, C#commit.disc_copies, R2),
-    do_update(Tid, disc_only_copies, C#commit.disc_only_copies, R3).
+    R4 = do_update(Tid, disc_only_copies, C#commit.disc_only_copies, R3),
+    do_update(Tid, external_copies, C#commit.external_copies, R4).
 
 %% Update the items
 do_update(Tid, Storage, [Op | Ops], OldRes) ->
@@ -1871,6 +1878,7 @@
 commit_nodes([C | Tail], AccD, AccR) 
         when C#commit.disc_copies == [],
              C#commit.disc_only_copies  == [],
+             C#commit.external_copies  == [],
              C#commit.schema_ops == [] ->
     commit_nodes(Tail, AccD, [C#commit.node | AccR]);
 commit_nodes([C | Tail], AccD, AccR) ->
@@ -1883,7 +1891,8 @@
     {D2, Tail2} = 
 	case C#commit.schema_ops of
 	    [] when C#commit.disc_copies == [],
-		    C#commit.disc_only_copies  == [] ->
+		    C#commit.disc_only_copies  == [],
+                    C#commit.external_copies  == [] ->
 		commit_decision(D, Tail, AccD, [N | AccR]);
 	    [] ->
 		commit_decision(D, Tail, [N | AccD], AccR);
