--- mnesia_controller.erl.orig	2009-01-14 13:22:53.000000000 -0800
+++ mnesia_controller.erl	2009-01-14 13:44:58.000000000 -0800
@@ -354,6 +354,8 @@
 	    do_force_load_table(Tab);
 	disc_only_copies ->
 	    do_force_load_table(Tab);
+        {external_copies, _} ->
+            do_force_load_table(Tab);        
 	unknown ->
 	    set({Tab, load_by_force}, true),
 	    cast({force_load_updated, Tab}),
@@ -1473,8 +1475,12 @@
 	    if   %% Avoid reading from disc_only_copies
 		NodeST == disc_only_copies ->
 		    ignore;
+		NodeST == external_copies ->
+		    ignore;
 		ReadST == disc_only_copies ->
 		    mnesia_lib:set_remote_where_to_read(Tab);
+		ReadST == external_copies ->
+		    mnesia_lib:set_remote_where_to_read(Tab);
 		true ->
 		    ignore
 	    end,
@@ -1523,6 +1529,7 @@
     Ram = Cs#cstruct.ram_copies,
     Disc = Cs#cstruct.disc_copies,
     DiscOnly = Cs#cstruct.disc_only_copies,
+    Ext = Cs#cstruct.external_copies,
     BetterCopies0 = mnesia_lib:remote_copy_holders(Cs) -- Downs,
     BetterCopies = BetterCopies0 -- Ram,
     AccessMode = Cs#cstruct.access_mode,
@@ -1555,7 +1562,7 @@
 	    false;
 	Storage == ram_copies ->
 	    if
-		Disc == [], DiscOnly == [] ->
+		Disc == [], DiscOnly == [], Ext == [] ->
 		    %% Nobody has copy on disc
 		    {true, {Tab, ram_only}};
 		true ->
@@ -1787,6 +1794,11 @@
 			dets:info(Tab, size), 
 			dets:info(Tab, file_size),
 			"bytes on disc");
+        {external_copies, Mod} ->
+            info_format(Tab,
+                        Mod:info(Tab, size),
+                        Mod:info(Tab, memory),
+                        "words of mem");
 	_ ->
 	    info_format(Tab, 
 			?ets_info(Tab, size),
