--- mnesia-4.4.2/ebin/mnesia.app	2008-04-07 12:34:06.000000000 -0600
+++ mnesia-4.4.2ext1/ebin/mnesia.app	2008-07-12 10:30:49.000000000 -0600
@@ -1,6 +1,6 @@
 {application, mnesia,
- [{description, "MNESIA  CXC 138 12"},
-  {vsn, "4.4.2"},
+ [{description, "MNESIA  CXC 138 12 ext1"},
+  {vsn, "4.4.2ext1"},
   {modules, [
 	     mnesia, 
 	     mnesia_backup, 
@@ -10,6 +10,7 @@
 	     mnesia_controller,
 	     mnesia_dumper, 
 	     mnesia_event, 
+             mnesia_ext,
 	     mnesia_frag, 
 	     mnesia_frag_hash, 
              mnesia_frag_old_hash,
